### DISCUSSION TITLE
How to add a horizontal scroll bar to table?

### DISCUSSION THREAD

**Discussion by jason-yau** [emilk/egui#1667] [Q&A]
I am creating table using egui_extra::TableBuilder now. The table is too big that I can not view all contents. I need a horizontal scroll bar to it. How do I add it?

**Comment by emilk** [comment]
Wrap it in `ScrollArea::horizontal`?

**Reply by jason-yau** [reply]
It does not work.

**Reply by Detector-I** [reply]
> Wrap it in `ScrollArea::horizontal`?

Old message but I think I need to answer it
I did test putting `TableBuilder` in `ScrollAres:: horizontal` and it did kinda worked, but problem was it moved vertical scroll all the way to the right side... So If I want to access it I need to scroll all the way to right... That's a problem.
One other thing that I tested was changing source of `TableBuilder` to use `ScrollArea::both` instead of default vertical one, but result was not what I expected, a horizontal scroll did appear, but when I try to scroll all the table and columns bugged (the vertical separator move incorrectly and...)

**Comment by ThePerfectComputer** [comment]
any updates on how to do this?

**Comment by omac777** [comment]
Hello, egui team,
Firstly, AWESOME WORK, really impressive stuff and thank you for all your efforts.

Mr. Jason Yau, I believe I answered your question in my sources.  Specifically this line:
https://gist.github.com/omac777/b094021a0a1128b8142eb34f20489ff3#file-app-rs-L184
```
ScrollArea::both().auto_shrink([false, false]).show_rows(
```
The "both()" means display a vscroll and hscroll(you were looking for hscroll).

The "auto_shrink" false false means do not auto shrink column width nor row height if nothing/somethingsmall is in a particular column item or row item. BUG: auto_shrink is still happening and sample code demonstrates that in the native gui.  I believe I also surfaced the same bug in the wasm target as well, but different behaviours as to where the column widths are changed/shrunk.

I did attempt an example table.  It's an instance of an egui eframe template tweaked with 18 column 96 row table with both horizontal and vertical scroll areas.  I would have attached a tgz but github is giving me issues uploading/tar/tgz/rs straight into the egui issues new issue page.  So here I am.
**This example works in both native and egui**, but results in different behaviours in both with respect to the number of cols/rows and the column widths for around columns 13 to 17.  Why?  I have no clue.  The fields inserted in each row are identical and are textedit controls.

I've provided 6 files through gist because uploading of files works there. app.rs, lib.rs, main.rs, Cargo.toml, index.html, and sw.js.
https://gist.github.com/omac777/b094021a0a1128b8142eb34f20489ff3
Before using these files, please 
```
git clone https://github.com/emilk/eframe_template egui_tst_tbl1
```
and then replace the respective files with the ones from the gist link above.

I did notice persistent storage affects the result updated web page/native gui,  I attempted to disable it so that the number of cols/rows is not fetched from the existing storage, but rather through the default().

To have a better chance at seeing what you would expect, the 1-2 are for native gui and 3-5 are for the wasm gui:
1. cargo clean
2. cargo build --release
3. trunk clean
4. trunk serve
5. clear your web browser cache settings

Please could you explain to me why the column widths change for the columns closer to the end of the row?  Thank you.

**Comment by omac777** [comment]
Using a monospace font would help, but it's not the source of the issue here.  The controls in the row are all the default textedit with its default width.  I believe the bug must be in the lower-levels of the plumbing, but I can't intuitively point out exactly where since the first 10 columns seem correctly aligned and at the end they seem to change.

**Comment by Detector-I** [comment]
Well any updates on this? I'm facing the same issue here and I'm not sure how to fix it...
having an option to add horizontal scroll to table widget would be good... Without it, it's not complete...

**Comment by druskus20** [comment]
Update: putting the table inside a ScrollArea works as of egui 0.23. 

There may be some issues toggling text-wrap on and off, because the scrollarea prevents the wrapping.
The workaround that I found that did the trick was to toggle Column::clip() on and off 

Something like: 

```rs
  let wrap_setting_changed =
      self.frontend_side.settings.wrap != self.previous_wrap_setting;
  let msg_col = Column::remainder()
      .at_least(200.0)
      .resizable(true)
      .clip(wrap_setting_changed); // <---
  table_builder = table_builder.column(msg_col); 
```
